src/tree/mod.rs

```diff
@@ -737,121 +752,73 @@ impl RatchetTree {
-        // Otherwise return the hash of the next parent
-        } else {
-            self.compute_parent_hash(parent)
-        };
-        // If the current node is a parent, replace the parent hash in that node
-        let current_node = &self.nodes[index.as_usize()];
-        if let Some(mut parent_node) = current_node.node.clone() {
-            parent_node.set_parent_hash(parent_hash);
-            self.nodes[index.as_usize()].node = Some(parent_node);
-            let updated_parent_node = &self.nodes[index.as_usize()];
-            updated_parent_node.hash(&self.ciphersuite).unwrap()
-        // Otherwise we reached the leaf level, just return the hash
-        } else {
-            parent_hash
-        }
+
+    /// Computes and sets the parent hashes in the tree. Returns the parent hash
+    /// for the ParentHashExtension.
+    pub fn set_parent_hashes(&mut self, index: LeafIndex) -> Vec<u8> {
+        compute_parent_hashes(self, index)
     }
-    /// Verifies the integrity of a public tree
-    pub fn verify_integrity(ciphersuite: &Ciphersuite, nodes: &[Option<Node>]) -> bool {
-        let node_count = NodeIndex::from(nodes.len());
-        let size = node_count;
-        for i in 0..node_count.as_usize() {
-            let node_option = &nodes[i];
-            if let Some(node) = node_option {
-                match node.node_type {
-                    NodeType::Parent => {
-                        let left_index = treemath::left(NodeIndex::from(i))
-                            .expect("verify_integrity: Error when computing left child of node.");
-                        let right_index = treemath::right(NodeIndex::from(i), size.into())
-                            .expect("verify_integrity: Error when computing right child of node.");
-                        if right_index >= node_count {
-                            return false;
+
+    /// Verify the parent hashes of the tree nodes. Returns `true` if all parent
+    /// hashes have successfully been verified and `false` otherwise.
+    pub fn verify_parent_hashes(&self) -> bool {
+        for (index, node) in self.nodes.iter().enumerate() {
+            // Only look at non-blank parent nodes
+            if index % 2 == 1 && node.is_full_parent() {
+                // Sanity check to see if it's a valid parent node
+                if node.node_type.is_leaf() || node.node.is_none() || node.key_package.is_some() {
+                    return false;
+                }
+                let left = treemath::left(index.into()).unwrap();
+                let right = treemath::right(index.into(), self.leaf_count()).unwrap();
+                // Unwrapping here is safe, because we know it is a full parent node
+                let parent_hash_field = &node.parent_hash().unwrap();
+                // Calculate the current
+                let current_hash =
+                    ParentHashInput::new(&self, index.into(), right.into(), parent_hash_field)
```

Comment:
Doesn't the hash we want to check against depend on which child we're looking at? I would expect it to be calculated ad-hoc when comparing with the `parent_hash` fields of the children as outlined [here](https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#section-7.4.2).

Reply:
As it currently stands, I think the procedure below does not quite match what's described in the spec.